
---
title: "Microbial community analysis using R and phyloseq"
author: "Florentin Constancias"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: true
    use_bookdown: true
    highlight: haddock
---


```{r packages, results=FALSE, warning=FALSE, include=FALSE}
require(tidyverse); packageVersion("tidyverse")
# require(phyloseq); packageVersion("phyloseq")
options(getClass.msg=FALSE) # https://github.com/epurdom/clusterExperiment/issues/66
```

This document is a guideline for community analysis using `R`, [Phyloseq](https://joey711.github.io/phyloseq/), [microbiome](http://microbiome.github.io/microbiome/), [vegan](https://cran.r-project.org/web/packages/vegan/vignettes/intro-vegan.pdf), [ggplot2](http://ggplot2.org/), [tidyverse](https://www.tidyverse.org/) and custom made functions to facilitate the analyses and visualization stored under the [DivComAnalyses](https://github.com/fconstancias/DivComAnalyses) repository.

We recommend using post-clustering approach for processing raw metabarcoding data and developed an `R` friendly pipeline named [metabaRpipe](https://github.com/fconstancias/metabaRpipe).


# Resources:

## Commnuity ecology
* [Lecture from Mahendra Mariadassou](https://www.gdc-docs.ethz.ch/MDA/handouts/MDA20_PhyloseqFormation_Mahendra_Mariadassou.pdf) including `phyloseq` R package (slides 4 - 61) and community ecology concepts (slides 62 to 156) 
* [GustaMe website](https://mb3is.megx.net/gustame)
* [Ramette A. Multivariate analyses in microbial ecology. FEMS Microbiol Ecol. 2007 Nov;62(2):142-60. doi: 10.1111/j.1574-6941.2007.00375.x. Epub 2007 Sep 20. PMID: 17892477; PMCID: PMC2121141.](https://pubmed.ncbi.nlm.nih.gov/17892477/)
* This tutorial.

## R:

* [installing R packages](https://r-coder.com/install-r-packages/)
* [R introduction](https://datacarpentry.org/R-ecology-lesson/01-intro-to-r.html)
* [R and the tidyverse](https://datacarpentry.org/R-ecology-lesson/03-dplyr.html)
* [R and the ggplot2 gramar of graphics](https://datacarpentry.org/R-ecology-lesson/04-visualization-ggplot2.html)


# Getting ready:

## Loading the necessary packages:

```{r}

```

## Loading the functions:

**internet connection needed**:
```{r sourcing, results=FALSE, warning=FALSE, include=TRUE}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_taxa_tests.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_normalisation.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_alpha.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_beta.R") 
```

## Exploring the example datasets:

### Data import:

We are going to illustrate some functions using two different datasets generated in the group and processed with the `metabaRpipe` pipeline. Those files are located.

Define phyloseq object's location: 

```{r}
ps1 = "../data-raw/ps_invivo.RDS"
```

Load phyloseq object:

```{r}
ps1 %>% # ps object wich is path of our physeq file on our computer
  readRDS() -> physeq_1  # the readRDS function and then it is directed/stored to a R object we named physeq
```

Call the phyloseq object:

```{r}
physeq_1
```


The dataset contains 3704 taxa, ASV and recorded in 116 samples `otu_table()` and 26 variables as metadata `sample_data()`. There are 7 levels in the taxonomic path of those ASV and `tax_table()` an ASV phylogenetic tree is stored `phy_tree()` in addition to the sequences `refseq()`

The different facets of the data can be easily access:

OTU/ASV table:

```{r}

physeq_1 %>%
  otu_table() %>%
  head() # print only the first few rows

```

N.B.: OTU (Operational Taxonomic Unit)can cover a lot of things including bands on a DGGE profile to OTU based on sequence similarity threshold or ASV. ASV (Amplicon Sequence Variant) is an type of OTU generated using post clustering approaches. See [here](https://www.nature.com/articles/ismej2017119), [here](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0264443) or [there](https://github.com/benjjneb/dada2/issues/62) for more details.

Check the tax table:

```{r}

physeq_1 %>%
  tax_table() %>%
  head() # print only the first few rows

```

Our phyloseq object has a 'Strain' taxonomic information we created - know taxonomical information at highest level + ASV id

```{r}

physeq_1 %>%
  tax_table() %>%
  head()# print only the first few rows

```

N.B.: I prefer to use tydiverse's pipe: %>% which makes the code more readable as compared to classical R.
For more info: <https://dplyr.tidyverse.org/> & <https://rstudio.com/resources/cheatsheets/>

Which is similar to:

```{r}
head(tax_table(physeq_1))
```

Check nucleotide sequences of the ASV:

```{r}
physeq_1 %>%
  refseq()
```


Check the metadata:

```{r}
physeq_1 %>%
  sample_variables()
```


```{r}
physeq_1 %>%
  sample_data() %>%
  head() # print only the first few rows
```

`Phyloseq` package has also some useful functions:

For instance to access sample names:

```{r}
physeq_1 %>%
  sample_names() %>%
  head() # print only the first few rows
```

which is similar to:

```{r}
physeq_1 %>%
  otu_table() %>%
  colnames() %>%
  head() # print only the first few rows
```

Some checks:

First top samples with highest highest total number of reads
```{r}
physeq_1 %>%
    sample_sums() %>%
    sort(decreasing = TRUE) %>%
    head()
```

First top ASV with highest highest total number of reads:

```{r}
physeq_1 %>%
    taxa_sums() %>%
    sort(decreasing = TRUE) %>%
    head()
```

Getting help on a particular function:

```{r}
?taxa_sums()
# then google
# then R package manual
# then ... colleague/ researchgate
```

### Manipulating a phyloseq object

Filtering, subseting samples or OTU based on metadata, taxonomy, abundance and prevalence is very easy in phyloseq:

### Filtering:

#### prune_taxa()/prune_samples() 

Two functions which prunes unwanted taxa/samples from a phyloseq object based on a vector of taxa to keep:

```{r}
#Create a logical vector answering the logical question: does the ASV represent more than 10 reads in total?

physeq_1 %>%
  taxa_sums() > 10 -> taxa_top

#taxa_top is indeed a logical vector

taxa_top %>%
  head()
```

```{r}
#filter the physeq object based on that vector

physeq_1 %>%
    prune_taxa(taxa = taxa_top) -> physeq_1_filtered
```

Note that all the data in our physeq_filtered object has been filtered: otu_table(), tax_table(), refseq(), ...
```{r}
physeq_1
```

```{r}
physeq_1_filtered
```

#### subset_taxa() subset_samples()

Those functions ubsets unwanted taxa/samples from a phyloseq object based on conditions that must be met:

Keep only OTU belonging to Firmicutes Phyla:

```{r}
physeq_1 %>%
  subset_taxa(Phylum == "Firmicutes")
```

Keep only samples from treatment A and no treatment to our metadata:

```{r}
physeq_1 %>%
  subset_samples(treatment_invivo %in% c("TreatA"))
```

Keep only samples from treatment A and no treatment to our metadata:

```{r}
physeq_1 %>%
  subset_samples(treatment_invivo %in% c("TreatA", "none"))
```

Keep only samples which are not defined by "none" in the "treatment_invivo" metadata:

```{r}
physeq_1 %>%
  subset_samples(treatment_invivo != "none") # != different from
```

Keep only samples which are not defined by "none" in the "treatment_invivo" metadata and with samples with BW values > 20:

```{r}
physeq_1 %>%
  subset_samples(treatment_invivo != "none" &  BW  > 20 ) # | = or, & = AND
```

### Smoothing taxonomical information:

####

```{r}
tax_table(physeq_1)[,c("Family","Genus","Species")] %>% 
  head()
```
`tax_glom() `agglomerates ASVs at a given taxonomic level. Finer taxonomic information is lost even if ASV names id is maintained.

Agglomerate taxonomy at the phylum level:

```{r}
physeq_1 %>%
  tax_glom(taxrank = "Family") %>% 
  tax_table() %>% 
  head()
```

```{r}
physeq_1 %>% 
  phyloseq_get_strains_fast()  -> physeq_1_st

physeq_1_st %>% 
  tax_table() %>% 
  head()
```

```{r}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_varia.R") 

physeq_most_abundant

physeq_get_unique

generate_color_palette

physeq_simplify_tax

physeq_glom_rename

```

```{r}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_normalisation.R") 

phyloseq_check_lib_size

phyloseq_rarefaction_curves

ggrare

phyloseq_rarefied_unrarefied_richness

phyloseq_filter_samples

phyloseq_filter_per_sample_OTU

phyloseq_remove_chloro_mitho

phyloseq_density_normalize
```

update metadata

```{r}

```

```{r}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_taxa_tests.R") 

phyloseq_run_DESeq2_pair_plots_formula

phyloseq_Maaslin2

phyloseq_boxplot_abundance

phyloseq_A_B_ratio

phyloseq_run_compare_means

phyloseq_run_Deseq

phyloseq_run_ALDEx2

phyloseq_correlate_taxa


```

```{r}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_alpha.R") 

phyloseq_alphas

plot_alphas

correlate_alpha

plot_taxa_abundances_over_time




```


```{r}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_alpha.R") 

phyloseq_alphas

plot_alphas

correlate_alpha

plot_taxa_abundances_over_time




```


```{r}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_beta.R") 

phyloseq_compute_bdiv

phyloseq_plot_bdiv

phyloseq_plot_PCoA_3d

physeq_pairwise_permanovas_adonis2

physeq_betadisper

phyloseq_TW

phyloseq_adonis_strata_perm

phyloseq_adonis

phyloseq_plot_ordinations_facet

phyloseq_ordinations_expl_var

phyloseq_distance_boxplot

phyloseq_add_taxa_vector

phyloseq_add_metadata_vector


phyloseq_dbRDA

phyloseq_pairwise_dbRDA

phyloseq_plot_dbrda

in_vitro_mIMT_STABvsTreat

phyloseq_generate_pcoa_per_variables




```







### Abundance counts:

#### rarefy_evendepth() down samples all samples to the same sequencing depth (a.k.a., Library Size) and prunes ASV that disappear from all samples as a result.

```{r}
physeq %>%
  rarefy_even_depth(rngseed = 123) %>% # important to specify rngseed: random number generator for reproducibility
  sample_sums() %>%
  head()
```

```{r}
physeq %>%
  sample_sums() %>%
  min()
```
In this example, we rarefied all the dataset to 23 sequences per sample because it is the minumum in the dataset.

Should you rarefy your data ? This question is still under 'debate': see  <https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531>
or <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5335496/>. It depend on the type of analysis you are going to do, the shape of your rarefaction curves, ... but to make it short it for alpha-diversity and beta diversity rarefaction is advised for standard analysis, not for negative bionomial based abundance testing.

#### transform_sample_counts applies a function to the abundance vector of each sample.  It is useful for normalization.

```{r}
physeq %>%
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  sample_sums() %>%
  head()
```

## Combining several operations:

```{r}
physeq %>%
  subset_samples(PCR_plate != "P3" | sample_number > 30) %>%
  prune_taxa(taxa =  taxa_sums(physeq) > 1000) %>%
  subset_taxa(Phylum != "unknown" & Genus %in% c("Lactococcus", "Streptococcus")) %>%
  filter_taxa(function(x) sum(x) > 0, TRUE) %>%
  rarefy_even_depth(sample.size = 1000, rngseed = 123)  %>%
  tax_glom(taxrank = "Genus") %>%
  transform_sample_counts(function(x) x/sum(x) * 100)
```




This website theme was created by modifiying the rmdformats readthedown format.

```{r}

sessionInfo()

```


# Alternative analysis approaches using user friendly platformes:

If you do not have experience with `R` and you do not want to invest time learning it or you want basic descritpions of your data, the easiest way to proceed is to use user friendly interfaces. 
There are several options:

## R-Shiny-interfaces:

These are `R` based interactive interfaces enabling a smooth code free experience.

  * [easy16S](https://shiny.migale.inrae.fr/app/easy16S/)

Select your data > RDS allows to directly load the phyloseq object.

  * [shiny-phyloseq](http://joey711.github.io/shiny-phyloseq/)


## QIIME2:

QIIMEâ„¢ (pronounced chime) stands for Quantitative Insights Into Microbial Ecology. It is a bioinformatic pipeline developed to ease the process of bioinformatic analysis of raw sequencing microbiome data - not explored here since we have processed the data using [metabaRpipe](https://github.com/fconstancias/metabaRpipe). It is also useful to visualize and analyze the processed data, in a user friendly experience:

### Generate `QIIME2` compatibles files from the phyloseq object:

Check the [metabaRpipe tutorial](https://github.com/fconstancias/metabaRpipe#7-export-qiime2-compatible-files).

```{bash, eval = FALSE}
Rscript metabaRpipe/Rscripts/phyloseq_export_qiime.Rscript \
-i dada2/physeq_phylo/phyloseq_phylo.RDS \
-o qiime2 \
-f metabaRpipe/Rscripts/functions.R 
```

Alternatively, this can also be run from R/Rstudio:

```{r, eval = FALSE}
require(tidyverse); source("https://raw.githubusercontent.com/fconstancias/metabaRpipe/master/Rscripts/functions.R")

readRDS("dada2/phyloseq.RDS") %>%  
  physeq_export_qiime(output_dir = "~/qiime2/")
```



This command will generate all the necessary processed files to analyze the metabarcoding data and as seen [here](https://github.com/fconstancias/metabaRpipe#exploring-the-outputs).

```{bash, eval = FALSE}
# List/ check the files are here:
ls
asv.fna			asv_neweek.qza		asv_rep_set.qza		qiime2_mapping_file.txt	qiime2_otu.qza		tax.txt
asv_biom.biom		asv_neweek.tre		qiime1_mapping_file.txt	qiime2_metadata.qzv	qiime2_taxonomy.qza
```

We have generated:

* A fasta file containing the sequence of the ASV: `asv.fna`. This corresponds to the `ps %>% refseq()` in `R`.
* A text file with the metadata: `qiime2_mapping_file.txt`. This corresponds to the `ps %>% sample_data()` in `R`.
* A text with the taxonomic path of the ASV: `tax.txt`. This corresponds to the `ps %>% sample_data()` in `R`.
* A biom file with the sample wise ASV counts: `asv_biom.biom`. This to the `ps %>% otu_table()` in `R`.
* A neweek ASV phylogenetic tree : `asv_neweek.tre`. This to the `ps %>% phy_tree()` in `R`.


### Import the data in `QIIME2`:


* Install `QIIME2` in a dedicated environment <https://docs.qiime2.org/2022.2/install/native/>

```{bash, eval = FALSE}
wget https://data.qiime2.org/distro/core/qiime2-2022.2-py38-osx-conda.yml
conda env create -n qiime2-2022.2 --file qiime2-2022.2-py38-osx-conda.yml
# OPTIONAL CLEANUP
rm qiime2-2022.2-py38-osx-conda.yml
# Activate the conda environment
conda activate qiime2-2022.2
# Test your installation
qiime --help
```

* Activate the environment change directory:
```{bash, eval = FALSE}
# Activate the qiime conda environment

conda activate qiime2-2022.2

# Navigate were the output of phyloseq_export_qiime.Rscript were generated:

cd qiime2/ 

# List/ check the files are here:
ls
asv.fna			asv_biom.biom		asv_neweek.tre		qiime1_mapping_file.txt	qiime2_mapping_file.txt	tax.txt
```

* Import the data in `qiime2` compatible format:

The following commands are required to import the files in `qza` `qiime2` specific format

```{bash, eval = FALSE}
# Import OTU/ASV table:
qiime tools import   \
--input-path asv_biom.biom \
--type 'FeatureTable[Frequency]' \
--input-format BIOMV100Format \
--output-path qiime2_otu.qza

# Import taxonomic table:
qiime tools import \
--type 'FeatureData[Taxonomy]' \
--input-format HeaderlessTSVTaxonomyFormat \
--input-path tax.txt \
--output-path qiime2_taxonomy.qza

# Import phylogenetic tree:
qiime tools import \
--input-path asv_neweek.tre \
--output-path asv_neweek.qza \
--type 'Phylogeny[Rooted]'

# Import ASV sequences:
qiime tools import \
--input-path asv.fna \
--output-path asv_rep_set.qza \
--type 'FeatureData[Sequence]'

# Import metadata:
qiime metadata tabulate \
--m-input-file qiime2_mapping_file.txt \
--o-visualization qiime2_metadata.qzv
```

The import successfully generated .qza files we will then use to visualize/ analyze the data:
```{bash, eval = FALSE}
# List/ check the files are here:
ls
asv.fna			asv_neweek.qza		asv_rep_set.qza		qiime2_mapping_file.txt	qiime2_otu.qza		tax.txt
asv_biom.biom		asv_neweek.tre		qiime1_mapping_file.txt	qiime2_metadata.qzv	qiime2_taxonomy.qza
```

### Diversity analyses:

### Data summary:

The first step is to summarize the ASV table using `qiime feature-table summarize`

```{bash, eval = FALSE}
qiime feature-table summarize \
--i-table qiime2_otu.qza \
--o-visualization  qiime2_otu.qzv \
--m-sample-metadata-file qiime2_mapping_file.txt
```

This will generate a `qiime2_otu.qzv` which we can open using:

```{bash, eval = FALSE}
qiime tools view qiime2_otu.qzv
```

In addition to the `Overview`, the `Interactive Sample Detail` summaries the number of Features (*i.e.*, here ASV) among samples. We can adjust the  `Sampling Depth` from the `Plot Controls` panel to see how many samples would be discarded when considering a minimum sampling depth cutoff or a rarefaction threshold.

An example of such summary can be found [here](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdocs.qiime2.org%2F2021.8%2Fdata%2Ftutorials%2Fmoving-pictures%2Ftable.qzv)

### Rarafaction analysis:

Another meaningful preliminary analysis is to perform rarefaction curves investigating the changes of alpha-diversity metrics (*e.g.*, Richness or Observed_features) as a function of the sequencing depth.

```{bash, eval = FALSE}
qiime diversity alpha-rarefaction \
  --i-phylogeny asv_neweek.qza \
  --i-table qiime2_otu.qza \
  --p-max-depth 5000 \
  --o-visualization CORE-METRICS-RESULTS/alpha-rarefaction.qzv
```    
    
```{bash, eval = FALSE}
qiime tools view CORE-METRICS-RESULTS/alpha-rarefaction.qzv
```

An example of rarefaction curves can be found [here](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdocs.qiime2.org%2F2021.8%2Fdata%2Ftutorials%2Fmoving-pictures%2Falpha-rarefaction.qzv). Ideally they should reach a plateau meaning the sequencing effort allows an exhaustive description of the diversity.

### Generating core-metrics for diversity analyses.

An important parameter here is the `--p-sampling-depth` which is the sequencing depth at which all the samples will be `rarefied` in order to avoid biases due to uneven sequencing depth. 

```{bash, eval = FALSE}
qiime diversity core-metrics-phylogenetic \
  --i-phylogeny asv_neweek.qza \
  --i-table qiime2_otu.qza \
  --p-sampling-depth 3500 \
  --m-metadata-file qiime2_mapping_file.txt \
  --output-dir CORE-METRICS-RESULTS
```

This command generates a lot of `qza` files corresponding to different `alpha` (i.e., faith_pd, observed_features, shannon, evenness) and `beta` diversity metrics (i.e., unweighted_unifrac, weighted_unifrac, jaccard_distance, bray_curtis_distance) and including PCoA ordniations for beta diversity metrics. (e.g., jaccard_pcoa_results.qza).

```{bash, eval = FALSE}
Saved FeatureTable[Frequency] to: CORE-METRICS-RESULTS/rarefied_table.qza
Saved SampleData[AlphaDiversity] to: CORE-METRICS-RESULTS/faith_pd_vector.qza
Saved SampleData[AlphaDiversity] to: CORE-METRICS-RESULTS/observed_features_vector.qza
Saved SampleData[AlphaDiversity] to: CORE-METRICS-RESULTS/shannon_vector.qza
Saved SampleData[AlphaDiversity] to: CORE-METRICS-RESULTS/evenness_vector.qza
Saved DistanceMatrix to: CORE-METRICS-RESULTS/unweighted_unifrac_distance_matrix.qza
Saved DistanceMatrix to: CORE-METRICS-RESULTS/weighted_unifrac_distance_matrix.qza
Saved DistanceMatrix to: CORE-METRICS-RESULTS/jaccard_distance_matrix.qza
Saved DistanceMatrix to: CORE-METRICS-RESULTS/bray_curtis_distance_matrix.qza
Saved PCoAResults to: CORE-METRICS-RESULTS/unweighted_unifrac_pcoa_results.qza
Saved PCoAResults to: CORE-METRICS-RESULTS/weighted_unifrac_pcoa_results.qza
Saved PCoAResults to: CORE-METRICS-RESULTS/jaccard_pcoa_results.qza
Saved PCoAResults to: CORE-METRICS-RESULTS/bray_curtis_pcoa_results.qza
Saved Visualization to: CORE-METRICS-RESULTS/unweighted_unifrac_emperor.qzv
Saved Visualization to: CORE-METRICS-RESULTS/weighted_unifrac_emperor.qzv
Saved Visualization to: CORE-METRICS-RESULTS/jaccard_emperor.qzv
Saved Visualization to: CORE-METRICS-RESULTS/bray_curtis_emperor.qzv
```

#### Alpha-diversity:

Alpha diversity is often defined as univariate within sample community characteristic, for more information please refer to slides 63, 66 to 79 [Mahendra Mariadassou's Lecture](https://www.gdc-docs.ethz.ch/MDA/handouts/MDA20_PhyloseqFormation_Mahendra_Mariadassou.pdf)
Now we have generated different metrics, we can visualize and explore associations with the metadata.

```{bash, eval = FALSE}
qiime diversity alpha-group-significance \
  --i-alpha-diversity CORE-METRICS-RESULTS/faith_pd_vector.qza \
  --m-metadata-file qiime2_mapping_file.txt \
  --o-visualization CORE-METRICS-RESULTS/faith-pd-group-significance.qzv

qiime diversity alpha-group-significance \
  --i-alpha-diversity CORE-METRICS-RESULTS/evenness_vector.qza \
  --m-metadata-file qiime2_mapping_file.txt \
  --o-visualization CORE-METRICS-RESULTS/evenness-group-significance.qzv
```

You will get errors related to the mapping file, we can quickly generate a new artificial mapping file linking samples to group A or B.
```{bash, eval = FALSE}
vim qiime2_mapping_file_2.txt
sampleid        group
R1F1-S66        A
R1F2-S300       A
R1F3-S90        B
Y2A15-2M-06-S78 B
Y2A15-2M-12-S77 B
Y3-R1F4-S136    B
```

Let's run the functions again:

```{bash, eval = FALSE}
qiime diversity alpha-group-significance \
  --i-alpha-diversity CORE-METRICS-RESULTS/faith_pd_vector.qza \
  --m-metadata-file qiime2_mapping_file_2.txt \
  --o-visualization CORE-METRICS-RESULTS/faith-pd-group-significance.qzv

qiime diversity alpha-group-significance \
  --i-alpha-diversity CORE-METRICS-RESULTS/evenness_vector.qza \
  --m-metadata-file qiime2_mapping_file_2.txt \
  --o-visualization CORE-METRICS-RESULTS/evenness-group-significance.qzv

```

And visualize the output, for instance for the evenness metric:

```{bash, eval = FALSE}
qiime tools view CORE-METRICS-RESULTS/evenness-group-significance.qzv
```

[Here]:(https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdocs.qiime2.org%2F2021.8%2Fdata%2Ftutorials%2Fmoving-pictures%2Fcore-metrics-results%2Ffaith-pd-group-significance.qzv) and example of the output.

If you want to explore the results for all the metrics you will have to open all the files and just considering one metadata covariate.


```{bash, eval = FALSE}
CORE-METRICS-RESULTS/evenness-group-significance.qzv				CORE-METRICS-RESULTS/unweighted-unifrac-group-significance.qzv
CORE-METRICS-RESULTS/faith-pd-group-significance.qzv
```

#### Beta-diversity:

Alpha diversity is often defined as multidimensional between sample community characteristic, for more information please refer to slides 63, 80 to 92 of [Mahendra Mariadassou's Lecture](https://www.gdc-docs.ethz.ch/MDA/handouts/MDA20_PhyloseqFormation_Mahendra_Mariadassou.pdf). This type of data is multidimensional and we need specific multidimensional approaches to visualize analyze such data.
Now we have generated different metrics, we can visualize and explore associations with the metadata.

##### PCoA ordination:

Principal Coordinate analysis is and unconstrained ordination aiming at exploring the main variation in the data. It is similar as a PCA but can deal with any type of metric.

Below are the commands to generate a PCoA using the `unweighted-unifrac` metric we generated using the `qiime diversity core-metrics-phylogenetic ` with the data rarefied at `3500` sequences per sample.

```{bash, eval = FALSE}
qiime emperor plot \
--i-pcoa CORE-METRICS-RESULTS/unweighted_unifrac_pcoa_results.qza \
--m-metadata-file qiime2_mapping_file_2.txt \
--o-visualization CORE-METRICS-RESULTS/unweighted-unifrac-emperor.qzv
```

We can then visualize the 3 first axes of the PCoA using the following command:

```{bash, eval = FALSE}
qiime tools view CORE-METRICS-RESULTS/unweighted-unifrac-emperor.qzv
```

An example of ordination plot can be found [here](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdocs.qiime2.org%2F2021.8%2Fdata%2Ftutorials%2Fmoving-pictures%2Fcore-metrics-results%2Funweighted_unifrac_emperor.qzv)

### Taxonomic visualisation:

The diversity analyses we have explored are not incorporating any taxonomic information and are considering ASV (i.e., features in `qiime2`). We can visualize community composition in terms of taxa proportion among samples using `qiime taxa barplot`:

```{bash, eval = FALSE}
qiime taxa barplot \
    --i-table qiime2_otu.qza \
    --i-taxonomy qiime2_taxonomy.qza \
    --m-metadata-file qiime2_mapping_file_2.txt \
    --o-visualization tax-bar-plots.qzv
```

The function generates a `qzv` plot we can explore using `qiime tools view `:

```{bash, eval = FALSE}
qiime tools view tax-bar-plots.qzv
```

An example of the output can be found [here](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdocs.qiime2.org%2F2021.8%2Fdata%2Ftutorials%2Fmoving-pictures%2Ftaxa-bar-plots.qzv)


### Filtering data:

There are many ways to filter the data: the samples, the ASV based on metadata, proportions, taxonomic information... Below an example on sample filtering based on metadata.

#### Based on metadata:

```{bash, eval = FALSE}
qiime feature-table filter-samples \
    --i-table qiime2_otu.qza \
    --m-metadata-file qiime2_mapping_file_2.txt \
    --p-where "[group]='B'" \
    --o-filtered-table qiime2_otu_groupB.qza
```

Analyzing the filtered dataset would require to start again with the beginning `qiime diversity core-metrics-phylogenetic` which will lead to loads of folders, outputs and makes it difficult to keep track of the code and analyses when things are getting a bit complex.


#### More [details](https://docs.qiime2.org/2022.2/tutorials/filtering/) about the filtering options.

A much more detailed `qiime2` tutorial can be found [here](https://curr-protoc-bioinformatics.qiime2.org/)


