---
title: ' Test dataset preparation `r Sys.Date()`'
author: "Florentin CONSTANCIAS"
date: "`r Sys.Date()`"
output:
  html_document: 
    toc: yes
    toc_depth: 4
    keep_md: yes
---
<style type="text/css">
div.main-container {
max-width: 1800px;
margin-left: auto;
margin-right: auto;
}
.Table-Normal {
position: relative;
//display: block;
margin: 10px auto;
padding: 0;
width: 100%;
height: auto;
border-collapse: collapse;
text-align: center;
}
</style>

```{r setup, include=FALSE}
rm(list = ls())

# knitr::opts_knit$set(root.dir = "/Users/fconstan/Projects/ETH/benoit/Soheil/New data/") 
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, results=FALSE, warning=FALSE, include=TRUE}
require(tidyverse); packageVersion("tidyverse")
require(phyloseq); packageVersion("phyloseq")
options(getClass.msg=FALSE) # https://github.com/epurdom/clusterExperiment/issues/66
```

# PolyFermS:

```{r, message = FALSE}
"~/Documents/GitHub/NRP72-FBT/data/processed/16S/ps_silva_dada2_human_chicken_meta.RDS" %>% 
  readRDS() %>% 
  subset_samples(Model == "Chicken" & Experiment == "Continuous" & Enrichment == "NotEnriched" & is.na(Paul) & Treatment %in% c("UNTREATED", "CTX", "VAN") & !Reactor %in% c("CR", "IR2")) %>%
  # subset_samples(tabled < 20000 & tabled > 1000) %>% 
  filter_taxa(function(x) sum(x > 0) > 0, TRUE) -> ps_poly

ps_poly %>% 
  sample_data() %>% 
  data.frame() %>% 
  select(-I7_Index_ID, -index, -I5_Index_ID, -index2, -metagenomic_sample_name,-raw_metagenomic_pairs,-Reactor_Treatment_Dose ,-Treatment_Dose,-Model, -Date, -HV292.1_Copy_Number_permL, -Enrichment, -CCUG59168_Copy_Number_permL,-CTX_Copy_Number_permL, -VAN_Copy_Number_permL, -Reactor_Treatment, -Sample_description) %>% 
  mutate(Treatment = recode(Treatment, 
                            `UNTREATED` = "Control",
                            `CTX` = "Antibiotic_A",
                            `VAN` = "Antibiotic_B"),
         Antibiotic = recode(Antibiotic, 
                             `UNTREATED` = "Control",
                             `CTX` = "Antibiotic_A",
                             `VAN` = "Antibiotic_B")) -> sample_data(ps_poly)

ps_poly %>% 
  sample_data() %>% 
  data.frame()
```  

```{r}
ps_poly %>% 
  filter_taxa(function(x) sum(x > 0) > 0, TRUE) %>% 
  saveRDS("~/Documents/GitHub/DivComAnalyses/data-raw/ps_PolyFermS.RDS")
```


# Invivo:
```{r, message = FALSE}

"~/Documents/GitHub/mIMT/data/physeq_silva132_clean_phangorn_tree_NAin.RDS" %>% 
  readRDS() %>% 
  subset_samples(sample_type %in% c("Feces","Cecum")) %>%
  subset_samples(!treatment %in% c("InoculumFermenter_slurry","InoculumFermenter", "Caecal_inoculum_F6")) %>%
  subset_samples(!is.na(treatment)) %>%
  subset_samples(!is.na(day)) %>%
  subset_samples(remove != "x") %>% #sample_data() %>% data.frame()
  subset_samples(tabled > 500) %>% 
  filter_taxa(function(x) sum(x > 0) > 0, TRUE) -> physeq
```  

```{r, message = TRUE, warning = TRUE}
sample_data(physeq)$treatment = factor(sample_data(physeq)$treatment,
                                       levels = c("H2O", "H2O_co",
                                                  "Eubiotic",
                                                  "Normobiotic",
                                                  "Dysbiotic",
                                                  "DSS", "DSS_co",
                                                  "Predni"))

sample_data(physeq)$sample_type = factor(sample_data(physeq)$sample_type,
                                         levels = c("Feces",
                                                    "Cecum"))

sample_data(physeq)$day = factor(sample_data(physeq)$day,
                                 levels = c("Day_-1",
                                            "Day_6",
                                            "Day_12",
                                            "Day_14"))

sample_data(physeq)$treatment_grouped <- factor(ifelse(sample_data(physeq)$treatment == "DSS_co", "DSS", 
                                                       ifelse(sample_data(physeq)$treatment == "H2O_co", "H2O", 
                                                              as.vector(sample_data(physeq)$treatment))),
                                                levels = c("H2O",
                                                           "Eubiotic",
                                                           "Normobiotic",
                                                           "Dysbiotic",
                                                           "DSS",
                                                           "Predni"))


sample_data(physeq)$treatment_invivo <- factor(ifelse(sample_data(physeq)$treatment_grouped %in% c("H2O") | sample_data(physeq)$day %in% c("Day_-1"), 
                                                      "none", 
                                                      as.vector(sample_data(physeq)$treatment_grouped)),
                                               levels = c("none",
                                                          "Eubiotic",
                                                          "Normobiotic",
                                                          "Dysbiotic",
                                                          "DSS",
                                                          "Predni"))

sample_data(physeq)$day_num <- as.numeric(parse_number(as.character(sample_data(physeq)$day)))

tax_table(physeq) <- tax_table(physeq)[,-8]


physeq %>% 
  sample_data() %>% 
  data.frame()
```  

```{r, message = TRUE, warning = TRUE}
physeq %>% 
  subset_samples(sample_type == "Feces" & 
                   experiment == "mIMT_2" &
                   treatment_grouped %in% c("H2O", 
                                            "Eubiotic",
                                            "Dysbiotic",
                                            "DSS")) -> ps_invivo

sample_names(ps_invivo) <- sample_data(ps_invivo)$Sample


ps_invivo %>% 
  sample_data() %>% 
  data.frame()
```  

```{r, message = TRUE, warning = TRUE}
ps_invivo %>% 
  sample_data() %>% 
  data.frame() %>% 
  select(-reads ,-remove,-experiment, -owner, -run, -region_1, -responsible_seq,-facility, -sample_type, -reactor, -day_IR, -day_fermentation, -treatment_period, -treatment_exvivo, -experiment_period, -period_treat, -IR_TR, -period_treat, -sample_file, -colon_lenght:-period_reactor, -sample_label, -Fermentation, ) %>% 
  select(Sample, everything()) %>% 
  mutate(treatment_invivo = recode(treatment_invivo, 
                                   `H2O` = "Control",
                                   `Eubiotic` = "TreatA",
                                   `Dysbiotic` = "TreatB",
                                   `DSS` = "TreatC"),
         treatment = recode(treatment, 
                            `H2O` = "Control",
                            `Eubiotic` = "TreatA",
                            `Dysbiotic` = "TreatB",
                            `DSS` = "TreatC"),
         treatment_grouped = recode(treatment_grouped, 
                                    `H2O` = "Control" ,                
                                    `Eubiotic` = "TreatA",
                                    `Dysbiotic` = "TreatB",
                                    `DSS` = "TreatC")) -> sample_data(ps_invivo)
ps_invivo %>% 
  sample_data() %>% 
  data.frame()
```
```{r}
ps_invivo %>% 
    filter_taxa(function(x) sum(x > 0) > 0, TRUE) %>% 
  saveRDS("~/Documents/GitHub/DivComAnalyses/data-raw/ps_invivo.RDS")
```
```{r, message = FALSE, warning= FALSE}
sessionInfo()
```
