---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# DivComAnalyses

<!-- badges: start -->
[![Travis build status](https://travis-ci.com/fconstancias/DivComAnalyses.svg?branch=master)](https://travis-ci.com/fconstancias/DivComAnalyses)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/fconstancias/DivComAnalyses?branch=master&svg=true)](https://ci.appveyor.com/project/fconstancias/DivComAnalyses)
[![Codecov test coverage](https://codecov.io/gh/fconstancias/DivComAnalyses/branch/master/graph/badge.svg)](https://codecov.io/gh/fconstancias/DivComAnalyses?branch=master)
<!-- badges: end -->

The goal of DivComAnalyses is to facilitate community analysis (e.g., 16S community metabarcoding, metaphlan taxonomic profiles from metagenomes). In data filtering (i.e., potential contaminants identified using decontam R package), help in normalizing data (i.e., 16S qPCR normalization), generate phylogenetic and taxonomic alpha-diversity metrics, visualize community composition and structure using ordination, facilitate statistical analysis based on DEseq2, ALDex, vegan R packages.

## Installation

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("fconstancias/DivComAnalyses")
```
## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(DivComAnalyses)
## basic example code
```

What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so:

```{r cars}
summary(cars)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub!
